avro/src/main/java/com/fasterxml/jackson/dataformat/avro/schema/RecordVisitor.java

```diff
@@ -190,6 +205,9 @@ public void optionalProperty(String name, JsonFormatVisitable handler,
             }
         }
         JsonNode defaultValue = parseJson(prop.getMetadata().getDefaultValue());
+        if(isDefaultsEnabled && defaultValue == null && !prop.getType().isPrimitive()) {
```

Comment:
Need to check if the writerSchema is a Union and that it contains NULL as a valid sub-schema in the union:

    writerSchema.getType() == Type.UNION && writerSchema.getIndexNamed(Type.NULL.getName()) != null

This can replace the isPrimitive check.

Reply:
Thanks, replaced condition based on proposal.